source/common/pool_impl.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+#include "nighthawk/common/exception.h"
+#include "nighthawk/common/poolable.h"
+
+#include "common/common/assert.h"
+#include "common/poolable_impl.h"
+
+namespace Nighthawk {
+
+template <typename Poolable> class PoolImpl {
+public:
+  using PoolDeletionDelegate = std::function<void(Poolable*)>;
+  using PoolInstanceConstructionDelegate = std::function<Poolable*(void)>;
+  using PoolInstanceResetDelegate = std::function<void(Poolable&)>;
+  using PoolablePtr = std::unique_ptr<Poolable, PoolDeletionDelegate>;
+
+  PoolImpl(PoolInstanceConstructionDelegate&& construction_delegate = nullptr,
+           PoolInstanceResetDelegate&& reset_delegate = nullptr)
+      : construction_delegate_(construction_delegate), reset_delegate_(reset_delegate) {}
+
+  ~PoolImpl() {
+    while (!pool_.empty()) {
+      Poolable* poolable = pool_.back().get();
+      all_.erase(std::remove(all_.begin(), all_.end(), poolable), all_.end());
+      pool_.pop_back();
+    }
+    // Inform the in-flight poolables that they are own their own now.
+    for (auto poolable : all_) {
+      poolable->mark_orphaned();
+    }
+  }
+
+  void addPoolable(std::unique_ptr<Poolable>&& poolable) {
+    ASSERT(poolable.get() != nullptr);
+    all_.push_back(poolable.get());
+    pool_.emplace_back(std::move(poolable));
+  }
+
+  PoolablePtr get() {
```

Comment:
This is kind of interesting, if I'm reading it correctly. It seems like the intended semantics are kind of like borrowing a book from the library: only one person can have the book at a time, and if the library burns down, the book belongs to you. But I have to wonder if it's really needed, and whether there's a simpler way to get what we need.

Reply:
Nice analogy, and correct. One TODO is that I still have to look into if we (still need) to use `dispatcher_.deferredDelete` here. It may be that we need to transfer ownership to the dispatcher instead of destructing directly here.